### PR TITLE
`Development`: Fix issues with LTI 1.3 for open edx

### PR DESCRIPTION
**Pull request by basak-akan**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

openEdx supports LTI 1.3 within their xblock-lti-consumer components. Each xblock-lti-consumer component acts as an external platform who consumes content (which means each xblock-lti-consumer has its own configuration details such as unique client id, authorization uri, jwkset uri, token uri). Also openEdx does not support dynamic configuration service (the service that you only provide a dynamic configuration link and all the configuration information communications done automatically). So Artemis administrators needs to manually fill in each xblock-lti-consumer components information into Artemis. 
Besides that since each xblock-lti-consumer components act as separate consumer it is impossible to keep track of which xblock-lti-consumer component launched an exercise in current setting. 

### Description

To adress above issues;

- new admin ui added to manually create lti platform
- added a new column to lti_resource_launch table which is a fk to the lti_platform_configuration id. In this way we can keep track of exercise launches and send the results back to correct xblock-lti-consumer component.

> [!WARNING]  
> **PLEASE ONLY DEPLOY TO TS11, CONTAINS DATABASE CHANGES!!!** 

### Steps for Testing

Prerequisites:
- 1 Student

1. Log into edX ( https://openedx.ase.cit.tum.de/ ) using one of the student credentials (i.e. artemis_test_user_1)

1. Navigate to LTI 1.3 Test Course
2. Select Introduction then Hello World section
3. Select launch exercise under Lti Consumer and then press ok
4. Verify iframe opens and nagivates to Artemis
5. Login artemis using same student credentials
6. Participate exercise and submit
7. Close iframe
8. Select progress on top bar
9. Verify you see exercise participation result

#### Performance Review
- [x] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [edit-lti-configuration.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6009/latest/artifact/shared/Coverage-Report-Client-Tests/app/admin/lti-configuration/edit-lti-configuration.component.ts.html) | 85% | ✅ |
| [lti-configuration.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6009/latest/artifact/shared/Coverage-Report-Client-Tests/app/admin/lti-configuration/lti-configuration.component.ts.html) | 94.73% | ✅ |
| [lti-configuration.service.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6009/latest/artifact/shared/Coverage-Report-Client-Tests/app/admin/lti-configuration/lti-configuration.service.ts.html) | 88.88% | ✅ |
| [course-lti-configuration.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6009/latest/artifact/shared/Coverage-Report-Client-Tests/app/course/manage/course-lti-configuration/course-lti-configuration.component.ts.html) | 96.77% | ✅ |
| [edit-course-lti-configuration.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6009/latest/artifact/shared/Coverage-Report-Client-Tests/app/course/manage/course-lti-configuration/edit-course-lti-configuration.component.ts.html) | 82.5% | ✅ |
| [course-management-tab-bar.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6009/latest/artifact/shared/Coverage-Report-Client-Tests/app/course/manage/course-management-tab-bar/course-management-tab-bar.component.ts.html) | 93.97% | ✅ |
| [course-management.service.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6009/latest/artifact/shared/Coverage-Report-Client-Tests/app/course/manage/course-management.service.ts.html) | 87.28% | ✅ |
| [online-course-configuration.model.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6009/latest/artifact/shared/Coverage-Report-Client-Tests/app/entities/online-course-configuration.model.ts.html) | 100% | ✅ |
| [lti-course-card.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6009/latest/artifact/shared/Coverage-Report-Client-Tests/app/lti/lti-course-card.component.ts.html) | 100% | ✅ |
| [lti13-exercise-launch.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6009/latest/artifact/shared/Coverage-Report-Client-Tests/app/lti/lti13-exercise-launch.component.ts.html) | 95.16% | ✅ |
| [lti13-select-course.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6009/latest/artifact/shared/Coverage-Report-Client-Tests/app/lti/lti13-select-course.component.ts.html) | 92.3% | ✅ |

#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [LtiPlatformConfiguration.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6009/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.domain/LtiPlatformConfiguration.html) | 92% | ✅ |
| [Lti13AgsClaim.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6009/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.domain.lti/Lti13AgsClaim.html) | 10% | ✅ ❌ |
| [Lti13ClientRegistration.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6009/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.domain.lti/Lti13ClientRegistration.html) | 84% | ✅ |
| [Lti13LaunchRequest.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6009/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.domain.lti/Lti13LaunchRequest.html) | 97% | ✅ |
| [LtiResourceLaunch.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6009/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.domain.lti/LtiResourceLaunch.html) | 76% | ✅ |
| [CourseRepository.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6009/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.repository/CourseRepository.html) | 76% | ✅ |
| [LtiPlatformConfigurationRepository.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6009/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.repository/LtiPlatformConfigurationRepository.html) | 76% | ✅ |
| [Lti13LaunchFilter.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6009/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.security.lti/Lti13LaunchFilter.html) | 100% | ✅ |
| [CourseService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6009/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service/CourseService.html) | 90% | ✅ |
| [OnlineCourseConfigurationService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6009/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service/OnlineCourseConfigurationService.html) | 95% | ✅ |
| [Lti13Service.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6009/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.connectors.lti/Lti13Service.html) | 88% | ✅ |
| [LtiService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6009/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.connectors.lti/LtiService.html) | 97% | ✅ |
| [CourseResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6009/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest/CourseResource.html) | 94% | ✅ |
| [LtiResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6009/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest/LtiResource.html) | 73% | ✅ |
| [AdminLtiConfigurationResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6009/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest.admin/AdminLtiConfigurationResource.html) | 86% | ✅ |
| [OnlineCourseDTO.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6009/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest.dto/OnlineCourseDTO.html) | 100% | ✅ |

### Screenshots

- edX instructor + artemis admin
1. edx instructor provides lti consumer information to artemis admin. 
2. Artemis admin creates new lti platform
3. edx instructor fills in artemis information into edx lti consumer.

https://github.com/ls1intum/Artemis/assets/11020241/0fb2b436-c0de-4601-93c1-80f6cce83271

- Student partipates into Artemis exercise via edX and see result on edX progress page

https://github.com/ls1intum/Artemis/assets/11020241/0041c0f7-e8f7-40d9-a43c-f7b1a768fc6f

## Summary by CodeRabbit

- **New Features**
    - Enhanced LTI 1.3 launch request handling with new client registration ID support.
    - Improved LTI platform configuration management, including the ability to add new configurations and retrieve them by ID.
    - Updated UI for LTI configuration in the admin section, including better handling of platform data and new nested routing for configuration editing.
    - Added a new button for creating LTI platform configurations, improving navigation and user experience.
- **Bug Fixes**
    - Updated conditional assignment for the `lineItem` variable to support various LMS storage locations.
- **Style**
    - Improved code readability by replacing specific import statements with wildcard imports in several Java files.
- **Tests**
    - Expanded test coverage with new tests for LTI platform configuration creation, online course retrieval for LTI dashboards, and service layer interactions.
    - Improved test reliability in the JavaScript component test by removing unnecessary `tick` function call and adding a spy for service method.

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/service/connectors/lti/Lti13Service.java [hunk 1]
```diff
@@ -172,10 +172,8 @@ private Lti13LaunchRequest launchRequestFrom(OidcIdToken ltiIdToken, String clie
     public void onNewResult(StudentParticipation participation) {
         Course course = courseRepository.findByIdWithEagerOnlineCourseConfigurationElseThrow(participation.getExercise().getCourseViaExerciseGroupOrCourseMember().getId());
 
-        LtiPlatformConfiguration ltiPlatformConfiguration = course.getOnlineCourseConfiguration().getLtiPlatformConfiguration();
-        ClientRegistration clientRegistration = onlineCourseConfigurationService.getClientRegistration(ltiPlatformConfiguration);
-        if (clientRegistration == null) {
-            log.error("Could not transmit score to external LMS for course {}: client registration not found", course.getTitle());
+        if (!course.isOnlineCourse()) {
+            log.error("Could not transmit score to external LMS for course {}:", course.getTitle());
```
[the remark this answers is not in the document]
Comment (basak-akan) [reply src/main/java/de/tum/in/www1/artemis/service/connectors/lti/Lti13Service.java:176]: Decided to maintain the current implementation of logging the error and continuing the execution flow. This approach ensures uninterrupted service, which is ok for our case

**Review by Gusti2010 — CHANGES_REQUESTED** [review]
Manually tested on ts11_Legacy. I don't know why, but i always get "No corresponding result available" when i try to submit. I even tried submitting the example solution. Do i have to create or upload a new exercise? (tested with test_user_1)

Screenshot 2024-02-10 at 14 41 56

**Comment by basak-akan** [comment]
> Manually tested on ts11_Legacy. I don't know why, but i always get "No corresponding result available" when i try to submit. I even tried submitting the example solution. Do i have to create or upload a new exercise? (tested with test_user_1)
> 
> Screenshot 2024-02-10 at 14 41 56

@Gusti2010 Thank you for testing my PR. This programming exercise issue is not related to this PR and needs admin actions for the test server. I already created and configured a new exercise (text exercise) and also updated the testing steps accordingly. I'd appreciate if you could check it out again.

**Review by reschandreas — DISMISSED** [review]
demonstrated in testing session, works as expected

**Review by Jan-Thurner — APPROVED** [review]
Tested in a TS and works as expected.

**Review by nityanandaz — DISMISSED** [review]
Testing session: could submit and see progress (user 4)

**Review by nata-81 — DISMISSED** [review]
Tested during testing session works well.

**Review by Kroko-fant — APPROVED** [review]
Tested on ts11 in a testing session

**Review by rstief — DISMISSED** [review]
Code looks good 👍

File: src/main/webapp/app/course/manage/course-lti-configuration/edit-course-lti-configuration.component.ts [hunk 2]
```diff
@@ -108,6 +108,11 @@ export class EditCourseLtiConfigurationComponent implements OnInit {
         });
     }
 
+    setPlatform(platform: LtiPlatformConfiguration) {
+        this.onlineCourseConfiguration.ltiPlatformConfiguration = platform;
+        this.onlineCourseConfigurationForm.get('ltiPlatformConfiguration')!.setValue(platform);
```
Comment (reschandreas) [inline src/main/webapp/app/course/manage/course-lti-configuration/edit-course-lti-configuration.component.ts:113]:
```suggestion
        this.onlineCourseConfigurationForm.get('ltiPlatformConfiguration')?.setValue(platform);
```

**Review by milljoniaer — APPROVED** [review]
LGTM

**Review by rstief — DISMISSED** [review]
💯

**Review by reschandreas — DISMISSED** [review]
reapprove

**Review by Jan-Thurner — DISMISSED** [review]
reapprove

**Review by nata-81 — APPROVED** [review]
reapprove

**Pull request merged**